ibis/common/tests/test_dispatch.py

```diff
@@ -0,0 +1,118 @@
+import collections
+import decimal
+
+from ibis.common.dispatch import lazy_singledispatch
+
+
+def test_lazy_singledispatch():
+    @lazy_singledispatch
```

Comment:
Is there any reason not to use `lazy_singledispatch` everywhere, other than that the eager version isn't maintained by us?

Reply:
This implementation actually looks cleaner than the one in functools. One major difference is that `functools.singledispatch()` supports type annotation, though that's something we could live without. I'd be curious of a `first dispatch` benchmark between the two.